### ISSUE TITLE
Support for ConvTranspose2d layers?

### ISSUE THREAD

**Issue by ErikBertolino** [tfjgeorge/nngeometry#23]
Hello!

Will there ever be support for ConvTranspose2d layers? Or is it a very easy corner case from Conv2D layers thjat can implement?

I tried to calculate the exact FIM for this autoencoder;

```
class GradConCAE(nn.Module):
    def __init__(self, in_channel=3):
        super(GradConCAE, self).__init__()

        self.down = nn.Sequential(
            nn.Conv2d(in_channel, 32, 4, stride=2, padding=2),
            nn.ReLU(),
            nn.Conv2d(32, 32, 4, stride=2, padding=2),
            nn.ReLU(),
            nn.Conv2d(32, 64, 4, stride=2, padding=2),
            nn.ReLU(),
            nn.Conv2d(64, 64, 4, stride=2, padding=2),
            nn.ReLU()
        )

        self.up = nn.Sequential(
            nn.ConvTranspose2d(64, 64, 4, stride=2, padding=2),  # output 4x4
            nn.ReLU(),
            nn.ConvTranspose2d(64, 32, 4, stride=2, padding=1),  # output 8x8
            nn.ReLU(),
            nn.ConvTranspose2d(32, 32, 4, stride=2, padding=2),  # output 14x14
            nn.ReLU(),
            nn.ConvTranspose2d(32, in_channel, 4, stride=2, padding=1),  # output 28x28
            nn.Sigmoid()
        )

        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        z = self.down(x)
        return self.up(z)
```

**Comment by tfjgeorge** [comment]
hey, the tricky part is to compute per-example gradients for the ConvTranspose2d layer, given the input of the layer, and the gradients w.r.t. the output (the preactivations) of the layer, similarly to what is done for conv layers here: https://github.com/tfjgeorge/nngeometry/blob/bdcf8a7afe5b9661b660d07ef5f5dd4baddec20c/nngeometry/utils.py#L6

I have to admit that I don't really know how to do this... I will have a closer look at it after the NeurIPS deadline!

**Comment by ErikBertolino** [comment]
Hey! Any update?

If it possible, can you guide how you did calculate it theoretically for the convolution layers? Is there any paper for this?

**Comment by tfjgeorge** [comment]
Hi, so I looked at what is needed to implement per_example gradients for transposed convolutions and unfortunately it is a bit more complicated.

For standard convolutions, I just used `unfold` which enables turning a convolution operation into a linear operation applied to unfolded input. From this I can just compute the per-example gradients. However with transposed convolution there is no such `unfold` function, which makes implementation more difficult.

Some references:
https://arxiv.org/abs/1912.06015 for a very good review of computing per example gradients for convolutions
https://github.com/vdumoulin/conv_arithmetic/blob/master/README.md for a visual explanation of what convolutions and transposed convolutions do, especially when varying the stride and padding arguments.

So unfortunately I won't be able to implement it myself. But if you write a function similar to `per_example_grad_conv` but applied to transposed convolutions I can integrate it in NNGeometry.

Best
